### DISCUSSION TITLE
How To Use max_pool In Flax

### DISCUSSION THREAD

**Discussion by nikhilanayak** [google/flax#1725] [Q&A] [answered]
You would like to see a new example being implemented by the Flax core team or the community? Please let us know by filling the following template.

### Description of the model to be implemented
I'm trying to use flax's `max_pool` function with a CNN.

### Dataset the model could be trained on
Using `jnp.zeros` should work, the example isn't about the data itself, more about the shape of the data. 

### Reference implementations in other frameworks
When I do `tf.keras.layers.MaxPooling3D(pool_size=(2,2,2))`, it scales down all 3 axes of the data by 2. I want to achieve a similar result using flax.

**Comment by marcvanzee** [comment, marked as answer]
*(Note: I moved this to a discussion since this doesn't really seem like a request for a new example but more like a question how to reproduce a TF op in Flax.)*

There's a bit of a subtle difference between the way [tf.nn.max_pool3d](https://www.tensorflow.org/api_docs/python/tf/nn/max_pool3d) and [flax.max_pool](https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.max_pool.html) work: If you only provide the `pool_size` in TF, this actually sets both the reduce window and the strides to that value. However, in Flax, the argument `window_shape` only sets the reduce window and not the strides. So if you want the same behavior you should set them both explicitely:

```python3
import tensorflow as tf
from flax import linen as nn
import numpy as np
tf.compat.v1.enable_eager_execution()

shape = (2, 30, 30, 30, 3)
reduce_window = (3, 3, 3)

inputs = numpy.random.random(shape)
tf_layer = tf.keras.layers.MaxPooling3D(pool_size=reduce_window)

tf_y = tf_layer(inputs).numpy()
flax_y = nn.max_pool(inputs, window_shape=reduce_window, strides=reduce_window)

np.testing.assert_allclose(tf_y, flax_y)
```

If you want to reduce over a window without resizing in TF, you should set `strides` to 1's explicitly:

```python
shape = (2, 30, 30, 30, 3)
reduce_window = (3, 3, 3)

inputs = numpy.random.random(shape)
tf_layer = tf.keras.layers.MaxPooling3D(pool_size=reduce_window, strides=(1, 1, 1))

tf_y = tf_layer(inputs).numpy()
flax_y = nn.max_pool(inputs, window_shape=reduce_window)

np.testing.assert_allclose(tf_y, flax_y)
```

**Reply by nikhilanayak** [reply]
Thanks, that worked perfectly!
